synapse/storage/user_directory.py

```diff
@@ -30,7 +28,162 @@
 logger = logging.getLogger(__name__)
 
 
-class UserDirectoryStore(SQLBaseStore):
+class UserDirectoryStore(BackgroundUpdateStore):
+    def __init__(self, db_conn, hs):
+        super(UserDirectoryStore, self).__init__(db_conn, hs)
+
+        self.server_name = hs.hostname
+
+        self.register_background_update_handler(
+            "populate_user_directory", self._populate_user_directory
+        )
+
+    @defer.inlineCallbacks
+    def _populate_user_directory(self, progress, batch_size):
+
+        TEMP_TABLE = "_temp_populate_user_directory_rooms"
+        state = self.hs.get_state_handler()
+
+        # If we don't have progress filed, delete everything.
+        if not progress:
+            yield self.delete_all_from_user_dir()
+
+        # Get all the rooms that we want to process.
+        def _make_staging_area(txn):
+            sql = (
+                "CREATE TABLE IF NOT EXISTS "
+                + TEMP_TABLE
+                + "(room_id TEXT NOT NULL, events BIGINT NOT NULL)"
+            )
+            txn.execute(sql)
+
+            # Get rooms we want to process from the database
+            sql = """
+                SELECT room_id, count(*) FROM current_state_events
+                GROUP BY room_id
+            """
+            txn.execute(sql)
+            rooms = [{"room_id": x[0], "events": x[1]} for x in txn.fetchall()]
+
+            self._simple_insert_many_txn(txn, TEMP_TABLE, rooms)
+
+            progress["stage"] = "processing"
+            self._background_update_progress_txn(
+                txn, "populate_user_directory", progress
+            )
+
+        def _delete_staging_area(txn):
+            sql = "DROP TABLE IF EXISTS " + TEMP_TABLE
+            txn.execute(sql)
+
+        if progress.get("stage") != "processing":
+            new_pos = yield self.get_max_stream_id_in_current_state_deltas()
+            progress["pos"] = new_pos
+            yield self.runInteraction(
+                "populate_user_directory_temp_build", _make_staging_area
+            )
+
+        def _get_next_batch(txn):
+            sql = """
+                SELECT room_id FROM %s
+                ORDER BY events DESC
+                LIMIT %s
+            """ % (
+                TEMP_TABLE,
+                str(batch_size),
+            )
+            txn.execute(sql)
+            rooms_to_work_on = txn.fetchall()
+
+            if not rooms_to_work_on:
+                return None
+
+            rooms_to_work_on = [x[0] for x in rooms_to_work_on]
+
+            sql = "SELECT COUNT(*) FROM " + TEMP_TABLE
+            txn.execute(sql)
+            progress["remaining"] = txn.fetchone()[0]
```

Comment:
If we're only doing this for logging below, there's probably not much reason to store it in `progress` at all

Reply:
it's kept there because it's the only shared state (since synapse can be restarted)